pyfr/integrators/writers.py

```diff
@@ -61,43 +63,106 @@ def _get_name_for_soln(self, etype, prank=None):
+            self._write = self.__writeParallel
+        else:
+            self._write = self.__writeSerial
+
+        if (parallel):
+            if rank == root:
+                solList = []
+                for mrank, meleinfo in enumerate(eleinfo):
+                    prank = self.rallocs.mprankmap[mrank]
+                    solList.extend(
+                            [
+                                (self._get_name_for_soln(etype, prank), dims)
+                                for (etype, dims) in meleinfo
+                            ])
+            else:
+                solList = None
+            self.solList = comm.bcast(solList, root=root)
+
+        else:
+            if rank == root:
+                self._mpi_rbufs = mpi_rbufs = []
+                self._mpi_rreqs = mpi_rreqs = []
+                self._mpi_names = mpi_names = []
+                self._loc_names = loc_names = []
+
+                for mrank, meleinfo in enumerate(eleinfo):
+                    prank = self.rallocs.mprankmap[mrank]
+                    for tag, (etype, dims) in enumerate(meleinfo):
+                        name = self._get_name_for_soln(etype, prank)
+
+                        if mrank == root:
+                            loc_names.append(name)
+                        else:
+                            rbuf = np.empty(dims, dtype=self.backend.fpdtype)
+                            rreq = comm.Recv_init(rbuf, mrank, tag)
+
+                            mpi_rbufs.append(rbuf)
+                            mpi_rreqs.append(rreq)
+                            mpi_names.append(name)
 
     def _write(self, path, solnmap, metadata):
+        raise Exception
+
+    def __writeParallel(self, path, solnmap, metadata):
+        from mpi4py import MPI
+
+        comm, rank, root = get_comm_rank_root()
+
+        _file = h5py.File(path, "w", driver='mpio', comm=comm)
+        sMap = {}
+        for s, shape in self.solList:
+            sMap[s] = _file.create_dataset(s, shape, dtype='float64')
+
+        for e in solnmap:
+            s = self._get_name_for_soln(e, self.rallocs.prank)
+            sMap[s][:] = solnmap[e]
+
+        if(rank == root):
+            mMap = []
+            for m in metadata:
+                if(isinstance(metadata[m], str)):
+                    mMap.append((m, (), 'S%d' % (len(metadata[m].encode()))))
+                else:
+                    raise NotImplementedError(
+                        "meta writer of type %s" % (str(type(metadata[m])))
+                                             )
+        else:
+            mMap = None
+
+        mMap = comm.bcast(mMap, root=root)
+        for name, shape, dtype in mMap:
+            d = _file.create_dataset(name, shape, dtype=dtype)
+            if(rank == root):
+                if(isinstance(metadata[m], str)):
+                    d.write_direct(np.array(metadata[name]).astype('S'))
+
+        _file.close()
+
+    def __writeSerial(self, path, solnmap, metadata):
```

Comment:
Unnecessary use of `__` (super-private) member.  The above three functions can be simplified by an assignment in `__init__`:

```
if parallel:
    self._writer = self._writer_parallel
else:
    self._writer = self._writer_serial
```